ILSpy/LoadedAssembly.cs

```diff
@@ -365,6 +376,31 @@ LoadResult LoadAssembly(Stream stream, PEStreamOptions streamOptions)
 			return new LoadResult(module);
 		}
 
+		LoadResult LoadCompressedAssembly(string filename)
+		{
+			const uint CompressedDataMagic = 0x5A4C4158; // Magic used for Xamarin compressed module header ('XALZ', little-endian)
+			using (var fileStream = new FileStream(fileName, FileMode.Open, FileAccess.Read))
+			using (var fileReader = new BinaryReader(fileStream))
+			{
+				// Read compressed file header
+				var magic = fileReader.ReadUInt32();
+				if (magic != CompressedDataMagic)
+					throw new InvalidDataException($"Xamarin compressed module header magic {magic} does not match expected {CompressedDataMagic}");
+				var descIdx = fileReader.ReadUInt32();
+				var uncLen = fileReader.ReadUInt32();
+				// fileReader stream position is now at compressed module data
+				var src = fileReader.ReadBytes((int)uncLen); // compressed length must be smaller than uncompressed length
```

Comment:
I would read the complete rest of the file instead of relying on an assumption here.

Reply:
Thanks, now using ``fileStream.Length`` to read all remaining data. Not accounting for compressed header length since ``ReadBytes()`` will stop at stream end anyway. We also save us from having a new constant and checking for  (theoretical case) ``fileStream.Length`` < 12